File: backend/app/core/domain/action_matrix.py
ACTION_MATRIX: dict[tuple[str, str], bool] = {
    ("new", "edit"): True,
    ("new", "delete"): True,
    ("new", "transition"): True,
    ("in_progress", "edit"): True,
    ("in_progress", "delete"): True,
    ("in_progress", "transition"): True,
    ("done", "edit"): False,
    ("done", "delete"): False,
    ("done", "transition"): False,
}

VALID_TRANSITIONS: dict[tuple[str, str], bool] = {
    ("new", "new"): True,
    ("new", "in_progress"): True,
    ("new", "done"): True,
    ("in_progress", "new"): True,
    ("in_progress", "in_progress"): True,
    ("in_progress", "done"): True,
    ("done", "done"): True,
}


def can_perform(state: str, action: str) -> bool:
    return ACTION_MATRIX.get((state, action), False)


def can_transition(from_state: str, to_state: str) -> bool:
    return VALID_TRANSITIONS.get((from_state, to_state), False)


def get_allowed_transitions(state: str) -> list[str]:
    return [to for (fr, to), ok in VALID_TRANSITIONS.items() if fr == state and ok]

File: frontend/src/features/tickets/actionMatrix.ts
import type { TicketStatus } from "../../api/types";

const ACTION_MATRIX: Record<string, Record<string, boolean>> = {
  new:         { edit: true, delete: true, transition: true },
  in_progress: { edit: true, delete: true, transition: true },
  done:        { edit: false, delete: false, transition: false },
};

const VALID_TRANSITIONS: Record<string, string[]> = {
  new:         ["new", "in_progress", "done"],
  in_progress: ["new", "in_progress", "done"],
  done:        ["done"],
};

export function canPerform(state: TicketStatus, action: string): boolean {
  return ACTION_MATRIX[state]?.[action] ?? false;
}

export function getAllowedTransitions(state: TicketStatus): TicketStatus[] {
  return (VALID_TRANSITIONS[state] ?? []) as TicketStatus[];
}
